File: pylith/problems/InitialConditionsPatch.py
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University of Chicago
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2015 University of California, Davis
#
# See COPYING for license information.
#
# ----------------------------------------------------------------------
#
# @file pylith/problems/InitialConditionsPatch.py
#
# @brief Python object for specifying initial conditions over a portion of the domain (patch).
#
# Factory: initial_conditions.

from pylith.problems.InitialConditions import InitialConditions
from .problems import InitialConditionsPatch as ModuleInitialConditions


def validateLabel(value):
    """
    Validate label for group/nodeset/pset.
    """
    if 0 == len(value):
        raise ValueError("Label for initial condition group/nodeset/pset in mesh not specified.")
    return value


class InitialConditionsPatch(InitialConditions, ModuleInitialConditions):
    """
    Python object for specifying initial conditions over a portion of the domain (patch).

    INVENTORY

    Properties
      - *label* Label marking subdomain.

    Facilities
      - *db* Spatial database with values for initial conditions.
    """

    import pyre.inventory

    label = pyre.inventory.str("label", default="", validator=validateLabel)
    label.meta["tip"] = "Label identifier for boundary."

    db = pyre.inventory.facility("db", default=SimpleDB, factory="spatial_database")
    db.meta["tip"] = "Spatial database with values for initial condition."

    # PUBLIC METHODS /////////////////////////////////////////////////////

    def __init__(self, name="initialconditionspatch"):
        """
        Constructor.
        """
        InitialConditions.__init__(self, name)
        return

    def preinitialize(self, mesh):
        """
        Setup initial conditions.
        """
        InitialConditions.preinitialize(self, mesh)

        ModuleInitialConditions.setMarkerLabel(self, self.label)
        ModuleInitialConditions.setDB(self, self.db)
        return

    def _configure(self):
        """
        Setup members using inventory.
        """
        InitialConditions._configure(self)
        return

    # PRIVATE METHODS ////////////////////////////////////////////////////

    def _createModuleObj(self):
        """
        Call constructor for module object for access to C++ object.
        """
        ModuleInitialConditionsPatch.__init__(self)


# FACTORIES ////////////////////////////////////////////////////////////

def initial_conditions():
    """
    Factory associated with InitialConditionsPatch.
    """
    return InitialConditionsPatch()


# End of file

File: libsrc/pylith/problems/InitialConditionsPatch.cc
// -*- C++ -*-
//
// ======================================================================
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University of Chicago
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2015 University of California, Davis
//
// See COPYING for license information.
//
// ======================================================================
//

#include <portinfo>

#include "InitialConditionsPatch.hh" // implementation of class methods

#include "pylith/topology/FieldQuery.hh" // USES FieldQuery

#include "spatialdata/spatialdb/SpatialDB.hh" // USES SpatialDB
#include "spatialdata/units/Nondimensional.hh" // USES Nondimensional

#include "pylith/utils/error.hh" // USES PYLITH_CHECK_ERROR
#include "pylith/utils/journals.hh" // USES PYLITH_COMPONENT_*
#include <cassert> // USES assert()

// ----------------------------------------------------------------------
// Constructor
pylith::problems::InitialConditionsPatch::InitialConditionsPatch(void) :
    _patchLabel(""),
    _db(NULL) {
    PyreComponent::setName("initialconditionspatch");
} // constructor


// ---------------------------------------------------------------------------------------------------------------------
// Destructor
pylith::problems::InitialConditionsPatch::~InitialConditionsPatch(void) {
    deallocate();
} // destructor


// ---------------------------------------------------------------------------------------------------------------------
// Deallocate PETSc and local data structures.
void
pylith::problems::InitialConditionsPatch::deallocate(void) {
    PYLITH_METHOD_BEGIN;

    _db = NULL; // :KLLUDGE: Should use shared pointer.

    PYLITH_METHOD_END;
} // deallocate


// ---------------------------------------------------------------------------------------------------------------------
// Set label for marker associated with patch.
void
pylith::problems::InitialConditionsPatch::setMarkerLabel(const char* value) {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("setMarkerLabel(value="<<value<<")");

    if (strlen(value) == 0) {
        throw std::runtime_error("Empty string given for initial conditions patch label.");
    } // if

    _patchLabel = value;

    PYLITH_METHOD_END;
} // setMarkerLabel


// ---------------------------------------------------------------------------------------------------------------------
// Get label marking constrained degrees of freedom.
const char*
pylith::problems::InitialConditionsPatch::getMarkerLabel(void) const {
    return _patchLabel.c_str();
} // getMarkerLabel


// ---------------------------------------------------------------------------------------------------------------------
// Set spatial database holding initial conditions.
void
pylith::problems::InitialConditionsPatch::setDB(spatialdata::spatialdb::SpatialDB* db) {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("setDB(db="<<db<<")");

    _db = db;

    PYLITH_METHOD_END;
} // setDB


// ---------------------------------------------------------------------------------------------------------------------
// Set solver type.
void
pylith::problems::InitialConditionsPatch::setValues(pylith::topology::Field* solution,
                                                    const spatialdata::units::Nondimensional& normalizer) {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("setValues(solution="<<solution<<", normalizer)");

    assert(solution);

    pylith::topology::FieldQuery fieldQuery(*solution);
    fieldQuery.initializeWithDefaultQueryFns();
    fieldQuery.setMarkerLabel(_patchLabel.c_str());
    fieldQuery.openDB(_db, normalizer.lengthScale());
    fieldQuery.queryDB();
    fieldQuery.closeDB(_db);

    PYLITH_METHOD_END;
} // setValues


// End of file
